lib/matplotlib/image.py

```diff
@@ -162,203 +210,158 @@ def changed(self):
-            ymin = ymin_old + iy0*dyintv/numrows
-            ymax = ymin_old + iy1*dyintv/numrows
-            dyintv = ymax - ymin
-            sy = dyintv/viewlim.height
-        else:
-            yslice = slice(0, numrows)
+        clipped_bbox = Bbox.intersection(out_bbox, clip_bbox)
 
-        if xslice != self._oldxslice or yslice != self._oldyslice:
-            self._imcache = None
-            self._oldxslice = xslice
-            self._oldyslice = yslice
+        if clipped_bbox is None:
+            return None, 0, 0, None
 
-        if self._imcache is None:
-            A = self._A
-            if self.origin == 'upper':
-                A = A[::-1]
+        out_width_base = clipped_bbox.width * magnification
+        out_height_base = clipped_bbox.height * magnification
 
-            if A.dtype == np.uint8 and A.ndim == 3:
-                im = _image.frombyte(A[yslice, xslice, :], 0)
-                im.is_grayscale = False
-            else:
-                if self._rgbacache is None:
-                    x = self.to_rgba(A, bytes=False)
-                    # Avoid side effects: to_rgba can return its argument
-                    # unchanged.
-                    if np.may_share_memory(x, A):
-                        x = x.copy()
-                    # premultiply the colors
-                    x[..., 0:3] *= x[..., 3:4]
-                    x = (x * 255).astype(np.uint8)
-                    self._rgbacache = x
-                else:
-                    x = self._rgbacache
-                im = _image.frombyte(x[yslice, xslice, :], 0)
-                if self._A.ndim == 2:
-                    im.is_grayscale = self.cmap.is_gray()
-                else:
-                    im.is_grayscale = False
-            self._imcache = im
+        if out_width_base == 0 or out_height_base == 0:
+            return None, 0, 0, None
+
+        if self.origin == 'upper':
+            # Flip the input image using a transform.  This avoids the
+            # problem with flipping the array, which results in a copy
+            # when it is converted to contiguous in the C wrapper
+            t0 = Affine2D().translate(0, -A.shape[0]).scale(1, -1)
         else:
-            im = self._imcache
+            t0 = IdentityTransform()
+
+        t0 += (
+            Affine2D()
+            .scale(
+                in_bbox.width / A.shape[1],
+                in_bbox.height / A.shape[0])
+            .translate(in_bbox.x0, in_bbox.y0)
+            + self.get_transform())
+
+        t = (t0
+            + Affine2D().translate(
+                -clipped_bbox.x0,
+                -clipped_bbox.y0)
+            .scale(magnification, magnification))
+
+        # So that the image is aligned with the edge of the axes, we want
+        # to round up the output width to the next integer.  This also
+        # means scaling the transform just slightly to account for the
+        # extra subpixel.
+        if (t.is_affine and round_to_pixel_border and
```

Comment:
`t` also includes `self.get_transform()` (on line 264), which could be any arbitrary transform.